src/jltypes.c

```diff
@@ -717,6 +717,84 @@ jl_value_t *simple_union(jl_value_t *a, jl_value_t *b)
     return tu;
 }
 
+int obviously_disjoint(jl_value_t *a, jl_value_t *b, int specificity);
+jl_value_t *simple_intersect(jl_value_t *a, jl_value_t *b, int overesi)
+{
+    // Unlike `Union`, we don't unwrap `UnionAll` here to avoid possible widening.
+    size_t nta = count_union_components(&a, 1, 0);
+    size_t ntb = count_union_components(&b, 1, 0);
+    size_t nt = nta + ntb;
+    jl_value_t **temp;
+    JL_GC_PUSHARGS(temp, nt+1);
+    size_t count = 0;
+    flatten_type_union(&a, 1, temp, &count, 0);
+    flatten_type_union(&b, 1, temp, &count, 0);
+    assert(count == nt);
+    size_t i, j;
+    // first remove disjoint elements.
+    for (i = 0; i < nt; i++) {
+        if (obviously_disjoint(temp[i], (i < nta ? b : a), 0))
+            temp[i] = NULL;
+    }
+    // then check subtyping.
+    int8_t *stemp = (int8_t *)alloca(count);
+    memset(stemp, 0, count);
+    for (i = 0; i < nta; i++) {
+        if (temp[i] == NULL) continue;
+        int hasfree = jl_has_free_typevars(temp[i]);
+        for (j = nta; j < nt; j++) {
+            if (temp[j] == NULL) continue;
+            int subs = simple_subtype2(temp[i], temp[j], hasfree || jl_has_free_typevars(temp[j]));
+            int subab = subs & 1, subba = subs >> 1;
+            if (subba && !subab) {
+                stemp[i] = -1;
+            }
+            if (subab && !subba) {
+                stemp[j] = -1;
+                temp[j] = temp[i];
```

Comment:
Does this make the algorithm slightly unstable to ordering? Eventually, for `!overesi`, this algorithm could return `Union{a, b}` for all `a` and `b` from the original `a` and `b` for which it could prove that element is a `<:` of the other union. That may often be `Union{}`, but doesn't have to be.

Secondly for `!overesi`, it similarly can return any type that is both `<:a` and `<:b` but `>:intersect(a,b)`. I don't think that is quite the same as removing any element from `b` that is potentially a subtype of an element `a`. For example, if it had `Tuple{E1}` and `Tuple{E2}` in `a`, but had `Tuple{Union{E1,E2}}` in `b`, it may be incorrect to "forget" that `Tuple{E2}` was in the union (unless `!overesi`)

Reply:
I'm afraid that I didn’t get the 1st concern, `simple_intersect`only joins preserved types in `a`, or preserved and substituted ones in`b`, thus it seems unlikely to return `Union{a,b}`?

But the 2nd concern is correct, the current algorithm works well only for injective elements. We’d better perform the “substitution” after the subtyping

Reply:
For `!overesi`, it isn't a concern, but an observation that this can return the Union of all provably preservable elements (all elements of `a` that are definite subtypes of `b` and all elements of `b` that are definitely subtypes of `a`)

Reply:
IIUC, this is exactly what we want?
The result should `>:` all elements from `a` if it `<:b` and all elements from `b` if it `<:a`.
I think there's no need to worry about `Union{Int, Real} ∩ Integer` (a[i] <: b[j] <: a[k]) as all other path should have normalize it.

Reply:
It may already be doing much of this, and I just haven't fully understood that part of the algorithm. The example I am thinking here is that if you had a=`Union{Int32, Int64, AbstractString}` and b=`Union{Integer, String}`, then the `!overesi` intersection can be any type that is `<:Union{Int32, Int64, String}` (often chosen as `Union{}` currently) while the `overesi` intersection is any value that is `>:Union{Int32, Int64, String}`, which might be chosen as either of `a` or `b` currently.

Reply:
For `!overesi`, the algorithm requires all undisjointable part in `a`(`b`) <: `b`(`a`), otherwise it would always return `Union{}`.
Since the "Subtitution" has been replaced with an elimination with partiality, elements in `a` would be removed if we can't prove it `<:b`, while elements in `b` would be removed only if we can prove it `>:a`.